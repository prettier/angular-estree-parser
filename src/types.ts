import * as ng from '@angular/compiler/src/expression_parser/ast';
import * as b from '@babel/types';

export interface NGBaseNode {
  type: string;
  start: number;
  end: number;
  loc: b.SourceLocation;
}

export type NGNode = { comments?: b.CommentLine[] } & (
  | b.Node
  | NGMicrosyntaxNode
  | NGEmptyExpression
  | NGPipeExpression
  | NGQuotedExpression
  | NGChainedExpression
);

export interface NGEmptyExpression extends NGBaseNode {
  type: 'NGEmptyExpression';
}

export interface NGPipeExpression extends NGBaseNode {
  type: 'NGPipeExpression';
  left: b.Expression;
  right: b.Identifier;
  arguments: b.Expression[];
}

export interface NGQuotedExpression extends NGBaseNode {
  type: 'NGQuotedExpression';
  prefix: string;
  value: string;
}

export interface NGChainedExpression extends NGBaseNode {
  type: 'NGChainedExpression';
  expressions: NGNode[];
}

export interface RawNGComment {
  type: 'Comment';
  value: string;
  span: RawNGSpan;
}

export interface RawNGSpan {
  start: number;
  end: number;
}

/**
 *     Microsyntax ::
 *       ( ( Exp | Let ) ( Sep? ( Let | As | KeyExp ) )* Sep? Key? )?
 *
 *     Sep ::
 *       ';'
 *       ','
 *
 *     Key ::
 *       StringLiteral
 *       Identifier ( '-' Identifier )*
 *
 *     Exp ::
 *       NGNode ( 'as' Key )?
 *
 *     KeyExp ::
 *       Key ':'? Exp
 *
 *     As ::
 *       Key 'as' Key
 *
 *     Let ::
 *       'let' Key ( '=' Key )?
 */
export interface NGMicrosyntax extends NGBaseNode {
  type: 'NGMicrosyntax';
  body: (
    | NGMicrosyntaxAs
    | NGMicrosyntaxExpression
    | NGMicrosyntaxKey
    | NGMicrosyntaxKeyedExpression
    | NGMicrosyntaxLet
  )[];
}

export type NGMicrosyntaxNode =
  | NGMicrosyntax
  | NGMicrosyntaxAs
  | NGMicrosyntaxExpression
  | NGMicrosyntaxKey
  | NGMicrosyntaxKeyedExpression
  | NGMicrosyntaxLet;

export interface NGMicrosyntaxKey extends NGBaseNode {
  type: 'NGMicrosyntaxKey';
  name: string;
}

export interface NGMicrosyntaxKeyedExpression extends NGBaseNode {
  type: 'NGMicrosyntaxKeyedExpression';
  key: NGMicrosyntaxKey;
  expression: NGMicrosyntaxExpression;
}

export interface NGMicrosyntaxExpression extends NGBaseNode {
  type: 'NGMicrosyntaxExpression';
  expression: NGNode;
  alias: NGMicrosyntaxKey | null;
}

export interface NGMicrosyntaxAs extends NGBaseNode {
  type: 'NGMicrosyntaxAs';
  key: NGMicrosyntaxKey;
  alias: NGMicrosyntaxKey;
}

export interface NGMicrosyntaxLet extends NGBaseNode {
  type: 'NGMicrosyntaxLet';
  key: NGMicrosyntaxKey;
  value: NGMicrosyntaxKey | null;
}
