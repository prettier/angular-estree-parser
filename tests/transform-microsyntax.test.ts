import * as b from '@babel/types';
import { parseTemplateBindings } from '../src/index.js';
import type { NGMicrosyntaxKeyedExpression } from '../src/types.js';
import { snapshotAst } from './helpers.js';

test.each`
  input                             | types
  ${''}                             | ${[]}
  ${' let hero '}                   | ${['NGMicrosyntaxLet']}
  ${' let hero = hello '}           | ${['NGMicrosyntaxLet']}
  ${' let hero of heroes '}         | ${['NGMicrosyntaxLet', 'NGMicrosyntaxKeyedExpression']}
  ${' let hero ; of : heroes '}     | ${['NGMicrosyntaxLet', 'NGMicrosyntaxKeyedExpression']}
  ${' as b '}                       | ${['NGMicrosyntaxAs']}
  ${' a '}                          | ${['NGMicrosyntaxExpression']}
  ${' a as b '}                     | ${['NGMicrosyntaxExpression']}
  ${' a , b '}                      | ${['NGMicrosyntaxExpression', 'NGMicrosyntaxKey']}
  ${' a ; b '}                      | ${['NGMicrosyntaxExpression', 'NGMicrosyntaxKey']}
  ${' a ; b c '}                    | ${['NGMicrosyntaxExpression', 'NGMicrosyntaxKeyedExpression']}
  ${' a ; b : c '}                  | ${['NGMicrosyntaxExpression', 'NGMicrosyntaxKeyedExpression']}
  ${' a ; b : c as d '}             | ${['NGMicrosyntaxExpression', 'NGMicrosyntaxKeyedExpression']}
  ${' a ; b as c '}                 | ${['NGMicrosyntaxExpression', 'NGMicrosyntaxAs']}
  ${' let "a" = "b" ; "c" as "d" '} | ${['NGMicrosyntaxLet', 'NGMicrosyntaxAs']}
  ${' let "\\"" '}                  | ${['NGMicrosyntaxLet']}
`('$input', ({ input, types }) => {
  const ast = parseTemplateBindings(input);
  expect(snapshotAst(ast, input)).toMatchSnapshot();
  expect(ast.body.map((node) => node.type)).toEqual(types);
});

test('Shorthand', () => {
  const code = 'someTmpl; context: {app}';
  const ast = parseTemplateBindings(code);
  const secondExpression = ast.body[1] as NGMicrosyntaxKeyedExpression;
  const objectExpression = secondExpression.expression
    .expression as b.ObjectExpression;
  const firstProperty = objectExpression.properties[0] as b.ObjectProperty;
  expect(firstProperty.shorthand).toBe(true);
});
