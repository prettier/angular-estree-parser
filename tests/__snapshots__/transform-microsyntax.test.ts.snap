// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
NGMicrosyntax { body: [] }
> 1 |
    | ^
`;

exports[` a  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression"] }
> 1 |  a 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a 
    |  ^
`;

exports[` a , b  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression", "NGMicrosyntaxKey"] }
> 1 |  a , b 
    |  ^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a , b 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a , b 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  a , b 
    |      ^
`;

exports[` a ; b  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression", "NGMicrosyntaxKey"] }
> 1 |  a ; b 
    |  ^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a ; b 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a ; b 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  a ; b 
    |      ^
`;

exports[` a ; b : c  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression", "NGMicrosyntaxKeyedExpression"] }
> 1 |  a ; b : c 
    |  ^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a ; b : c 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a ; b : c 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxKeyedExpression { key: "NGMicrosyntaxKey", expression: "NGMicrosyntaxExpression" }
> 1 |  a ; b : c 
    |      ^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  a ; b : c 
    |      ^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a ; b : c 
    |          ^
--------------------------------------------------------------------------------
Identifier { name: "c" }
> 1 |  a ; b : c 
    |          ^
`;

exports[` a ; b : c as d  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression", "NGMicrosyntaxKeyedExpression"] }
> 1 |  a ; b : c as d 
    |  ^^^^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a ; b : c as d 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a ; b : c as d 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxKeyedExpression { key: "NGMicrosyntaxKey", expression: "NGMicrosyntaxExpression" }
> 1 |  a ; b : c as d 
    |      ^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  a ; b : c as d 
    |      ^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: "NGMicrosyntaxKey" }
> 1 |  a ; b : c as d 
    |          ^^^^^^
--------------------------------------------------------------------------------
Identifier { name: "c" }
> 1 |  a ; b : c as d 
    |          ^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "d" }
> 1 |  a ; b : c as d 
    |               ^
`;

exports[` a ; b as c  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression", "NGMicrosyntaxAs"] }
> 1 |  a ; b as c 
    |  ^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a ; b as c 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a ; b as c 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxAs { key: "NGMicrosyntaxKey", alias: "NGMicrosyntaxKey" }
> 1 |  a ; b as c 
    |      ^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  a ; b as c 
    |      ^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "c" }
> 1 |  a ; b as c 
    |           ^
`;

exports[` a ; b c  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression", "NGMicrosyntaxKeyedExpression"] }
> 1 |  a ; b c 
    |  ^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a ; b c 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a ; b c 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxKeyedExpression { key: "NGMicrosyntaxKey", expression: "NGMicrosyntaxExpression" }
> 1 |  a ; b c 
    |      ^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  a ; b c 
    |      ^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  a ; b c 
    |        ^
--------------------------------------------------------------------------------
Identifier { name: "c" }
> 1 |  a ; b c 
    |        ^
`;

exports[` a as b  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxExpression"] }
> 1 |  a as b 
    |  ^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: "NGMicrosyntaxKey" }
> 1 |  a as b 
    |  ^^^^^^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a as b 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  a as b 
    |       ^
`;

exports[` as b  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxAs"] }
> 1 |  as b 
    |  ^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxAs { key: "NGMicrosyntaxKey", alias: "NGMicrosyntaxKey" }
> 1 |  as b 
    |  ^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "$implicit" }
> 1 |  as b 
    |  ^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  as b 
    |     ^
`;

exports[` let "\\""  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxLet"] }
> 1 |  let "\\"" 
    |  ^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxLet { key: "NGMicrosyntaxKey", value: null }
> 1 |  let "\\"" 
    |  ^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: '"' }
> 1 |  let "\\"" 
    |      ^^^^
`;

exports[` let "a" = "b" ; "c" as "d"  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxLet", "NGMicrosyntaxAs"] }
> 1 |  let "a" = "b" ; "c" as "d" 
    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxLet { key: "NGMicrosyntaxKey", value: "NGMicrosyntaxKey" }
> 1 |  let "a" = "b" ; "c" as "d" 
    |  ^^^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "a" }
> 1 |  let "a" = "b" ; "c" as "d" 
    |      ^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "b" }
> 1 |  let "a" = "b" ; "c" as "d" 
    |            ^^^
--------------------------------------------------------------------------------
NGMicrosyntaxAs { key: "NGMicrosyntaxKey", alias: "NGMicrosyntaxKey" }
> 1 |  let "a" = "b" ; "c" as "d" 
    |                  ^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "c" }
> 1 |  let "a" = "b" ; "c" as "d" 
    |                  ^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "d" }
> 1 |  let "a" = "b" ; "c" as "d" 
    |                         ^^^
`;

exports[` let hero  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxLet"] }
> 1 |  let hero 
    |  ^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxLet { key: "NGMicrosyntaxKey", value: null }
> 1 |  let hero 
    |  ^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "hero" }
> 1 |  let hero 
    |      ^^^^
`;

exports[` let hero ; of : heroes  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxLet", "NGMicrosyntaxKeyedExpression"] }
> 1 |  let hero ; of : heroes 
    |  ^^^^^^^^^^^^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxLet { key: "NGMicrosyntaxKey", value: null }
> 1 |  let hero ; of : heroes 
    |  ^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "hero" }
> 1 |  let hero ; of : heroes 
    |      ^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKeyedExpression { key: "NGMicrosyntaxKey", expression: "NGMicrosyntaxExpression" }
> 1 |  let hero ; of : heroes 
    |             ^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "of" }
> 1 |  let hero ; of : heroes 
    |             ^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  let hero ; of : heroes 
    |                  ^^^^^^
--------------------------------------------------------------------------------
Identifier { name: "heroes" }
> 1 |  let hero ; of : heroes 
    |                  ^^^^^^
`;

exports[` let hero = hello  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxLet"] }
> 1 |  let hero = hello 
    |  ^^^^^^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxLet { key: "NGMicrosyntaxKey", value: "NGMicrosyntaxKey" }
> 1 |  let hero = hello 
    |  ^^^^^^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "hero" }
> 1 |  let hero = hello 
    |      ^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "hello" }
> 1 |  let hero = hello 
    |             ^^^^^
`;

exports[` let hero of heroes  1`] = `
NGMicrosyntax { body: ["NGMicrosyntaxLet", "NGMicrosyntaxKeyedExpression"] }
> 1 |  let hero of heroes 
    |  ^^^^^^^^^^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxLet { key: "NGMicrosyntaxKey", value: null }
> 1 |  let hero of heroes 
    |  ^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "hero" }
> 1 |  let hero of heroes 
    |      ^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKeyedExpression { key: "NGMicrosyntaxKey", expression: "NGMicrosyntaxExpression" }
> 1 |  let hero of heroes 
    |           ^^^^^^^^^
--------------------------------------------------------------------------------
NGMicrosyntaxKey { name: "of" }
> 1 |  let hero of heroes 
    |           ^^
--------------------------------------------------------------------------------
NGMicrosyntaxExpression { expression: "Identifier", alias: null }
> 1 |  let hero of heroes 
    |              ^^^^^^
--------------------------------------------------------------------------------
Identifier { name: "heroes" }
> 1 |  let hero of heroes 
    |              ^^^^^^
`;
